src/server/pkg/work/work.go

```diff
@@ -117,106 +199,124 @@ type ProcessFunc func(context.Context, *Task, *Task) error
+		return err
+	}
+	for {
+		select {
+		case e := <-claimWatch.Watch():
+			if e.Type == watch.EventError {
+				return e.Err
+			}
+			if e.Type != watch.EventDelete {
+				continue
+			}
+			var subtaskKey string
+			if err := e.Unmarshal(&subtaskKey, &Claim{}); err != nil {
 				return err
 			}
-			if taskInfo.State != State_RUNNING {
-				cancel()
-				return errutil.ErrBreak
+			if err := taskEntry.runSubtask(w.subtaskFunc(task, subtaskKey, processFunc)); err != nil {
+				return err
 			}
-			return nil
-		})
-	})
-	// Process the subtasks.
-	eg.Go(func() (retErr error) {
-		// (bryce) cleanup of subtasks needs to be re-thought.
-		// Cannot cleanup in this manner, because it messes up subtasks claims when a worker errors.
-		// Cleaning up in the master results with a race condition with workers
-		// waiting on work termination signal and subtask claims to expire.
-		//defer func() {
-		//	// Attempt to cleanup subtask entries.
-		//	if _, err := col.NewSTM(context.Background(), w.etcdClient, func(stm col.STM) error {
-		//		w.subtaskCol.ReadWrite(stm).DeleteAllPrefix(work.Id)
-		//		return nil
-		//	}); err != nil && retErr == nil {
-		//		retErr = err
-		//	}
-		//}()
-		if err := w.processSubtasks(ctx, task); err != nil {
-			return err
-		}
-		// Wait for a deletion event (ttl expired) before attempting to process the subtasks again.
-		return w.subtaskCol.ReadOnly(ctx).WatchOneF(task.Id, func(e *watch.Event) error {
-			return w.processSubtasks(ctx, task)
-		}, watch.WithFilterPut())
-	})
-	if err := eg.Wait(); err != nil {
-		// Work terminated.
-		if ctx.Err() == context.Canceled {
-			return nil
+		case e := <-subtaskWatch.Watch():
+			if e.Type == watch.EventError {
+				return e.Err
+			}
+			var subtaskKey string
+			if err := e.Unmarshal(&subtaskKey, &TaskInfo{}); err != nil {
+				return err
+			}
+			if err := taskEntry.runSubtask(w.subtaskFunc(task, subtaskKey, processFunc)); err != nil {
+				return err
+			}
+		case <-taskEntry.ctx.Done():
+			return taskEntry.ctx.Err()
 		}
-		return err
 	}
-	return nil
 }
 
-func (w *Worker) processSubtasks(ctx context.Context, task *Task) error {
-	for _, subtask := range task.Subtasks {
-		subtaskKey := path.Join(task.Id, subtask.Id)
+func (w *Worker) subtaskFunc(task *Task, subtaskKey string, processFunc ProcessFunc) subtaskFunc {
+	return func(ctx context.Context) error {
+		// (bryce) this should be refactored to have the check and claim in the same stm.
+		// there is a rare race condition that does not affect correctness, but it is less
+		// than ideal because a subtask could get run once more than necessary.
 		subtaskInfo := &TaskInfo{}
-		if err := w.subtaskCol.Claim(ctx, subtaskKey, subtaskInfo, func(ctx context.Context) error {
-			if err := w.processFunc(ctx, task, subtask); err != nil {
+		if _, err := col.NewSTM(ctx, w.etcdClient, func(stm col.STM) error {
+			return w.subtaskCol.ReadWrite(stm).Get(subtaskKey, subtaskInfo)
+		}); err != nil {
+			return err
+		}
```

Comment:
Could you put the read inside the `Claim` block?  Or is it common for workers to claim tasks that are already completed?

Reply:
If you notice the comment above this, I would like to refactor this so that the read and claim happen in the same stm. Although this is not really critical, it does prevent a rare race condition that does not affect correctness, but could result with duplicate work (although I prevent updating the subtask entry when this occurs). Basically, we could have a worker see that a subtask is running, then attempt to claim it immediately after a different worker has finished it and the claim expired (so the original worker thinks it has not been processed and does not see a claim on it). I added a check after a subtask completes to see if the subtask is already completed, which will result with the worker that did the duplicate work to not update the entry (we want the results from when it was originally processed).